<a href="https://colab.research.google.com/github/taaha3244/AtomCampAIBootcamp/blob/main/EDA_project_cricket_world_cup_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib as mpl
import os
from sklearn.impute import SimpleImputer


# Ignore all warnings
import warnings
warnings.filterwarnings('ignore')
warnings.filterwarnings(action='ignore', category=DeprecationWarning)
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.2f' % x)
#pd.set_option('display.max_rows', None)

Loading all the datasets

In [ ]:
file_paths = [
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/innings.csv",
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/matches.csv",
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/overBallDetails.csv",
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/overHistory.csv",
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/players.csv",
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/teams.csv",
    "C:/Users/taaha/Downloads/Cricket_Dataset/ICC Cricket World Cup/venues.csv"
    # File paths for the files
]

In [ ]:
dataframes = []  # List to store dataframes

for file_path in file_paths:
    df = pd.read_csv(file_path)
    dataframes.append(df)


Assigning Each Dataframe its name

In [ ]:
innings=dataframes[0]
matches=dataframes[1]
overBallDetails=dataframes[2]
overHistory=dataframes[3]
players=dataframes[4]
teams=dataframes[5]
venues=dataframes[6]


Lets look at the shape of all the DataFrames

In [ ]:
for df in dataframes:
    print(f"The shape of the dataframe is: ",df.shape)

The shape of the dataframe is:  (2682, 18)
The shape of the dataframe is:  (1700, 174)
The shape of the dataframe is:  (157323, 8)
The shape of the dataframe is:  (110411, 18)
The shape of the dataframe is:  (28911, 9)
The shape of the dataframe is:  (354, 6)
The shape of the dataframe is:  (193, 5)


The max number of rows in our data are 157323 and max columns are 174
Lets investigate 174 columns thats a lot of columns.They are in matches

In [ ]:
matches.head()

id  venueId                     toss.elected  toss.winner  team1Id  \
0  2859       73      Australia, who chose to bat    Australia     2808   
1  2860       40    New Zealand, who chose to bat  New Zealand     2810   
2  2861       71        England, who chose to bat      England     2812   
3  2862       39  West Indies, who chose to field  West Indies     2814   
4  2863       82    Sri Lanka, who chose to field    Sri Lanka     2816   

   team2Id                 matchDate              matchEndDate  \
0     2809  1975-06-07T11:00:00+0000  1975-06-07T23:00:00+0000   
1     2811  1975-06-07T11:00:00+0000  1975-06-07T23:00:00+0000   
2     2813  1975-06-07T11:00:00+0000  1975-06-07T23:00:00+0000   
3     2815  1975-06-07T11:00:00+0000  1975-06-07T23:00:00+0000   
4     2817  1975-06-11T11:00:00+0000  1975-06-11T23:00:00+0000   

      matchDateMs  matchEndDateMs matchStatus_outcome  \
0 171371000000.00 171414000000.00                   A   
1 171371000000.00 171414000000.00                   B   
2 171371000000.00 171414000000.00                   A   
3 171371000000.00 171414000000.00                   B   
4 171716000000.00 171760000000.00                   A   

               matchStatus_text  matchStatus_victoryMarginRuns  \
0      Australia won by 73 runs                            NaN   
1   New Zealand won by 181 runs                            NaN   
2       England won by 202 runs                            NaN   
3  West Indies won by 9 wickets                            NaN   
4      Australia won by 52 runs                            NaN   

   matchStatus_victoryMarginWickets  matchStatus_victoryMarginInningsRuns  \
0                               NaN                                   NaN   
1                               NaN                                   NaN   
2                               NaN                                   NaN   
3                               NaN                                   NaN   
4                               NaN                                   NaN   

  description matchSummary matchType            tournamentLabel  oversLimit  \
0     Match 1          NaN       ODI  Prudential World Cup 1975          60   
1     Match 2          NaN       ODI  Prudential World Cup 1975          60   
2     Match 3          NaN       ODI  Prudential World Cup 1975          60   
3     Match 4          NaN       ODI  Prudential World Cup 1975          60   
4     Match 5          NaN       ODI  Prudential World Cup 1975          60   

   totalBalls  isLimitedOvers match.summary  matchStatus umpire.name.5  \
0         NaN            True           NaN          NaN           NaN   
1         NaN            True           NaN          NaN           NaN   
2         NaN            True           NaN          NaN           NaN   
3         NaN            True           NaN          NaN           NaN   
4         NaN            True           NaN          NaN           NaN   

   Unnamed: 25  Unnamed: 26  Unnamed: 27  Unnamed: 28  Unnamed: 29  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 30  Unnamed: 31  Unnamed: 32  Unnamed: 33  Unnamed: 34  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 35  Unnamed: 36  Unnamed: 37  Unnamed: 38  Unnamed: 39  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          Na

Seems like columns after Umpire name are all empty .Lets drop these extra columns

In [ ]:
matches= matches.drop(matches.iloc[:, 25:],axis = 1)

In [ ]:
print(f"The Shape of Matches is : ",matches.shape)

The Shape of Matches is :  (1700, 25)


Investigating Innings Data

In [ ]:
innings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2682 entries, 0 to 2681
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             2682 non-null   int64  
 1   matchId        2682 non-null   int64  
 2   inningsNumber  2682 non-null   int64  
 3   runRate        2682 non-null   float64
 4   overProgress   2682 non-null   float64
 5   runs           2682 non-null   int64  
 6   wkts           2682 non-null   int64  
 7   ballsFaced     2682 non-null   int64  
 8   fours          2682 non-null   int64  
 9   sixes          2682 non-null   int64  
 10  allOut         2682 non-null   bool   
 11  noBallRuns     2682 non-null   int64  
 12  wideRuns       2682 non-null   int64  
 13  byeRuns        2682 non-null   int64  
 14  legByeRuns     2682 non-null   int64  
 15  penaltyRuns    2682 non-null   int64  
 16  battingTeamId  601 non-null    float64
 17  bowlingTeamId  601 non-null    float64
dtypes: bool(

Total 2682 rows and 18 columns .Has missing values in battingTeamId and bowlingTeamId

In [ ]:
#Lets focus on the missing values first
innings.isnull().sum()

id                  0
matchId             0
inningsNumber       0
runRate             0
overProgress        0
runs                0
wkts                0
ballsFaced          0
fours               0
sixes               0
allOut              0
noBallRuns          0
wideRuns            0
byeRuns             0
legByeRuns          0
penaltyRuns         0
battingTeamId    2081
bowlingTeamId    2081
dtype: int64

'battingTeamId' and 'bowlingTeamId' are Secondary keys for teams Dataframe. And 2081 of them are missing. That is 76% of the innings dataframe. Surely we cant delete the missing values rows. For, now lets replace it with 'Missing_id'

In [ ]:
innings[['battingTeamId', 'bowlingTeamId']] = innings[['battingTeamId', 'bowlingTeamId']].fillna('Mising_id')

Looking for duplicates

In [ ]:
innings[innings.duplicated()]

Empty DataFrame
Columns: [id, matchId, inningsNumber, runRate, overProgress, runs, wkts, ballsFaced, fours, sixes, allOut, noBallRuns, wideRuns, byeRuns, legByeRuns, penaltyRuns, battingTeamId, bowlingTeamId]
Index: []

In [ ]:
def calculate_z_scores(df):
    z_scores=(df-df.mean(numeric_only=True))/df.std(numeric_only=True)
    outliers=(z_scores>2) | (z_scores<-2)
    return outliers.sum()

In [ ]:
calculate_z_scores(innings)

allOut             0
ballsFaced       332
battingTeamId      0
bowlingTeamId      0
byeRuns          114
fours            215
id                 0
inningsNumber      0
legByeRuns       164
matchId            0
noBallRuns       162
overProgress     171
penaltyRuns       14
runRate          106
runs             109
sixes            177
wideRuns         105
wkts              83
dtype: int64

There are some outliers in the dataset,will further look for them in visualization

Lets look at the Matches DataFrame

In [ ]:
matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1700 entries, 0 to 1699
Data columns (total 25 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   id                                    1700 non-null   int64  
 1   venueId                               1700 non-null   int64  
 2   toss.elected                          1600 non-null   object 
 3   toss.winner                           1600 non-null   object 
 4   team1Id                               1700 non-null   int64  
 5   team2Id                               1700 non-null   int64  
 6   matchDate                             1700 non-null   object 
 7   matchEndDate                          1674 non-null   object 
 8   matchDateMs                           1675 non-null   float64
 9   matchEndDateMs                        1358 non-null   float64
 10  matchStatus_outcome                   1689 non-null   object 
 11  matchStatus_text 

It looks like alot of columns have missing values

In [ ]:
def cat_num(df):
    # Separate categorical and numerical columns
    categorical_columns = df.select_dtypes(include=['object']).columns
    numerical_columns = df.select_dtypes(exclude=['object']).columns

    # Display the separated columns
    print("Categorical Columns:")
    print(categorical_columns)
    print("\nMissing Values are : ")
    print(df[categorical_columns].isnull().sum())

    print("\nNumerical Columns:")
    print(numerical_columns)
    print("\nMissing Values are : ")
    print(df[numerical_columns].isnull().sum())


In [ ]:
cat_num(matches)

Categorical Columns:
Index(['toss.elected', 'toss.winner', 'matchDate', 'matchEndDate',
       'matchStatus_outcome', 'matchStatus_text', 'description',
       'matchSummary', 'matchType', 'tournamentLabel', 'match.summary',
       'umpire.name.5'],
      dtype='object')

Missing Values are : 
toss.elected            100
toss.winner             100
matchDate                 0
matchEndDate             26
matchStatus_outcome      11
matchStatus_text          8
description               0
matchSummary            799
matchType                 0
tournamentLabel           0
match.summary           822
umpire.name.5          1698
dtype: int64

Numerical Columns:
Index(['id', 'venueId', 'team1Id', 'team2Id', 'matchDateMs', 'matchEndDateMs',
       'matchStatus_victoryMarginRuns', 'matchStatus_victoryMarginWickets',
       'matchStatus_victoryMarginInningsRuns', 'oversLimit', 'totalBalls',
       'isLimitedOvers', 'matchStatus'],
      dtype='object')

Missing Values are : 
id                  

Handling missing values in categorical columns involves making decisions about how to impute or handle these missing entries. Here are some common strategies:

1. Imputation with Mode:
Approach: Replace missing values with the most frequent category (mode) in the column.

2. Create a New Category for Missing Values:
Approach: Introduce a new category to represent missing values.

Adding Most frequent valus in these categorical columns will not be viable because of high cadinality as there are various unique values in our dataset

Segregating Categorical And Numerical Columns

In [ ]:
categorical_columns = matches.select_dtypes(include=['object']).columns
numerical_columns = matches.select_dtypes(exclude=['object']).columns

There are some date columns on Categorical Columns list so lets dig in to these date columns

In [ ]:
matches.loc[:,['matchEndDate','matchDate']].head(15)

matchEndDate                 matchDate
0   1975-06-07T23:00:00+0000  1975-06-07T11:00:00+0000
1   1975-06-07T23:00:00+0000  1975-06-07T11:00:00+0000
2   1975-06-07T23:00:00+0000  1975-06-07T11:00:00+0000
3   1975-06-07T23:00:00+0000  1975-06-07T11:00:00+0000
4   1975-06-11T23:00:00+0000  1975-06-11T11:00:00+0000
5   1975-06-11T23:00:00+0000  1975-06-11T11:00:00+0000
6   1975-06-11T23:00:00+0000  1975-06-11T11:00:00+0000
7   1975-06-11T23:00:00+0000  1975-06-11T11:00:00+0000
8   1975-06-14T23:00:00+0000  1975-06-14T11:00:00+0000
9   1975-06-14T23:00:00+0000  1975-06-14T11:00:00+0000
10  1975-06-14T23:00:00+0000  1975-06-14T11:00:00+0000
11  1975-06-14T23:00:00+0000  1975-06-14T11:00:00+0000
12  1975-06-18T23:00:00+0000  1975-06-18T11:00:00+0000
13  1975-06-18T23:00:00+0000  1975-06-18T11:00:00+0000
14  1975-06-21T23:00:00+0000  1975-06-21T11:00:00+0000

Seems like match Start and End dates are same in One day World Cups. so, missing values in match end date should be imputed from match start date

In [ ]:
matches['matchEndDate']=matches['matchEndDate'].fillna(matches['matchDate'])

Imputing rest of the categorical columns as  "MISSING"

In [ ]:
matches_cat=matches[categorical_columns].fillna('Missing')

In [ ]:
matches[numerical_columns].isnull().sum()

id                                         0
venueId                                    0
team1Id                                    0
team2Id                                    0
matchDateMs                               25
matchEndDateMs                           342
matchStatus_victoryMarginRuns           1554
matchStatus_victoryMarginWickets        1557
matchStatus_victoryMarginInningsRuns    1700
oversLimit                                 0
totalBalls                              1383
isLimitedOvers                             0
matchStatus                             1700
dtype: int64

'matchDateMs' and 'matchEndMs' are derived out of match date columns and are in UNIX datetime format. These columns can be dropped and if this date time format is req. in analysis we can convert it from date columns

The same is the case with 'matchStatus_victoryMarginRuns','matchStatus_victoryMarginWickets',
matchStatus_victoryMarginInningsRuns' and 'MatchStatus'.Derived from other columns from innings table and having more than 80% of data missing.    

In [ ]:
matches_num=matches[numerical_columns].drop(['matchDateMs','matchEndDateMs','matchStatus_victoryMarginRuns','matchStatus_victoryMarginWickets','matchStatus_victoryMarginInningsRuns','matchStatus'],axis=1)

In [ ]:
matches_num

id  venueId  team1Id  team2Id  oversLimit  totalBalls  \
0       2859       73     2808     2809          60         NaN   
1       2860       40     2810     2811          60         NaN   
2       2861       71     2812     2813          60         NaN   
3       2862       39     2814     2815          60         NaN   
4       2863       82     2816     2817          60         NaN   
...      ...      ...      ...      ...         ...         ...   
1695  102108       62       19       15          50      300.00   
1696  102843   100013       20       14          50      300.00   
1697  101735       72       11       16          50      300.00   
1698  102844   100013       14       13          50      300.00   
1699  102109       54       19       15          50      300.00   

      isLimitedOvers  
0               True  
1               True  
2               True  
3               True  
4               True  
...              ...  
1695            True  
1696            True  
1697            True  
1698            True  
1699            True  

[1700 rows x 7 columns]

total Balls column has approx 1400 missing values , it can be filled with 300 but from 1975 to 1983 in three world cups overs limit is 60 so 360 balls

from Datetime column Extracting year and appending it as another column

In [ ]:
# Convert 'Date' column to datetime format
matches_cat['matchDate'] = pd.to_datetime(matches_cat['matchDate'],format='ISO8601',utc=True)
matches_cat['matchDate']
# # Extract the year and append it to the DataFrame
matches_cat['Year'] = matches_cat['matchDate'].dt.year


# # Display the DataFrame with the new 'Year' column
# print(matches_cat)

In [ ]:
matches=pd.concat([matches_num,matches_cat],axis=1)

In [ ]:
matches.isnull().sum()

id                        0
venueId                   0
team1Id                   0
team2Id                   0
oversLimit                0
totalBalls             1383
isLimitedOvers            0
toss.elected              0
toss.winner               0
matchDate                 0
matchEndDate              0
matchStatus_outcome       0
matchStatus_text          0
description               0
matchSummary              0
matchType                 0
tournamentLabel           0
match.summary             0
umpire.name.5             0
Year                      0
dtype: int64

In [ ]:
condition_1 = matches['Year'].isin([1975, 1978, 1974])
condition_2 = ~condition_1

# Define values to fill based on conditions
value_condition_1 = 360
value_condition_2 = 300

# Use conditions with fillna
matches['totalBalls']=matches['totalBalls'].fillna(pd.Series(np.where(condition_1, value_condition_1,value_condition_2)))

In [ ]:
matches.isnull().sum()

id                     0
venueId                0
team1Id                0
team2Id                0
oversLimit             0
totalBalls             0
isLimitedOvers         0
toss.elected           0
toss.winner            0
matchDate              0
matchEndDate           0
matchStatus_outcome    0
matchStatus_text       0
description            0
matchSummary           0
matchType              0
tournamentLabel        0
match.summary          0
umpire.name.5          0
Year                   0
dtype: int64

In [ ]:
matches.describe()

id   venueId       team1Id       team2Id  oversLimit  totalBalls  \
count   1700.00   1700.00       1700.00       1700.00     1700.00     1700.00   
mean   22892.42   4913.96   13677682.80   15417866.26       50.26      300.53   
std    35844.45  21293.21  136935065.21  145322817.11        2.19        5.61   
min      543.00      1.00          0.00          0.00       21.00      300.00   
25%     2970.75     45.75       1569.25       1569.75       50.00      300.00   
50%     4328.50     84.00       3271.00       3272.00       50.00      300.00   
75%    21565.25    128.00       8300.00       8301.00       50.00      300.00   
max   102980.00 100236.00 2147212060.00 2089042683.00       60.00      360.00   

         Year  
count 1700.00  
mean  2013.76  
std     10.11  
min   1975.00  
25%   2013.00  
50%   2017.00  
75%   2020.00  
max   2023.00

In [ ]:
overBallDetails.set_index('id')

countingBall  nonCountingBall score  bowlerId  facingBatsmanId  \
id                                                                       
1                  1                1     1   5591.00          1220.00   
2                  2                2     .   5591.00          1601.00   
3                  3                3     .   5591.00          1601.00   
4                  4                4   1Wd   5591.00          1601.00   
5                  4                5     .   5591.00          1601.00   
...              ...              ...   ...       ...              ...   
157319             5                5     .   1894.00           857.00   
157320             6                6     .   1894.00           857.00   
157321             1                1     4   4567.00         17118.00   
157322             2                2     .   4567.00         17118.00   
157323             3                3     W   4567.00         17118.00   

        boundary  overHistoryId  
id                               
1          False          84666  
2          False          84666  
3          False          84666  
4          False          84666  
5          False          84666  
...          ...            ...  
157319     False         110410  
157320     False         110410  
157321      True         110411  
157322     False         110411  
157323     False         110411  

[157323 rows x 7 columns]

In [ ]:
overBallDetails.isnull().sum()

id                   0
countingBall         0
nonCountingBall      0
score                0
bowlerId           101
facingBatsmanId      3
boundary             0
overHistoryId        0
dtype: int64

In this DataFrame bowlerId has missing values which will be imputed as missing as other IDs in different columns

In [ ]:
overBallDetails[['bowlerId','facingBatsmanId']]=overBallDetails[['bowlerId','facingBatsmanId']].fillna('Missing_id')

In [ ]:
overBallDetails.isnull().sum()

id                 0
countingBall       0
nonCountingBall    0
score              0
bowlerId           0
facingBatsmanId    0
boundary           0
overHistoryId      0
dtype: int64

# overHistory

In [ ]:
overHistory.head()

ovNo ovBalls/0 ovBalls/1 ovBalls/2 ovBalls/3 ovBalls/4 ovBalls/5 ovBalls/6  \
0     1         1         .         .         .       1Wd         .         1   
1     2         1         .         .         .         1         1       NaN   
2     3         1         W         1         .       1Nb         .         .   
3     4         4       1Nb         .         .         .         4         .   
4     5         .         .         4         .         .         .       NaN   

   id  inningId ovBalls/7 ovBalls/8 ovBalls/9 ovBalls/10 ovBalls/11  \
0   1         1       NaN       NaN       NaN        NaN        NaN   
1   2         1       NaN       NaN       NaN        NaN        NaN   
2   3         1       NaN       NaN       NaN        NaN        NaN   
3   4         1       NaN       NaN       NaN        NaN        NaN   
4   5         1       NaN       NaN       NaN        NaN        NaN   

  ovBalls/12 ovBalls/13  bowlerId  
0        NaN        NaN       NaN  
1        NaN        NaN       NaN  
2        NaN        NaN       NaN  
3        NaN        NaN       NaN  
4        NaN        NaN       NaN

In [ ]:
overHistory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110411 entries, 0 to 110410
Data columns (total 18 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   ovNo        110411 non-null  int64  
 1   ovBalls/0   110411 non-null  object 
 2   ovBalls/1   110164 non-null  object 
 3   ovBalls/2   109906 non-null  object 
 4   ovBalls/3   109665 non-null  object 
 5   ovBalls/4   109405 non-null  object 
 6   ovBalls/5   108996 non-null  object 
 7   ovBalls/6   14563 non-null   object 
 8   id          110411 non-null  int64  
 9   inningId    110411 non-null  int64  
 10  ovBalls/7   1998 non-null    object 
 11  ovBalls/8   278 non-null     object 
 12  ovBalls/9   45 non-null      object 
 13  ovBalls/10  9 non-null       object 
 14  ovBalls/11  2 non-null       object 
 15  ovBalls/12  1 non-null       object 
 16  ovBalls/13  1 non-null       object 
 17  bowlerId    25729 non-null   float64
dtypes: float64(1), int64(3), object(14)
memory u

In [ ]:
overHistory.isnull().sum()

ovNo               0
ovBalls/0          0
ovBalls/1        247
ovBalls/2        505
ovBalls/3        746
ovBalls/4       1006
ovBalls/5       1415
ovBalls/6      95848
id                 0
inningId           0
ovBalls/7     108413
ovBalls/8     110133
ovBalls/9     110366
ovBalls/10    110402
ovBalls/11    110409
ovBalls/12    110410
ovBalls/13    110410
bowlerId       84682
dtype: int64

OverBalls7 to OverBalls 13 shows that there were cases when more than 6 Ballas were Bowled due to Wides or Maybe Noballs
Imputing First 6 balls missing values with Dot ball as it is the most probable outcome out of all. Will use Mode to do this

In [ ]:
overHistory.columns

Index(['ovNo', 'ovBalls/0', 'ovBalls/1', 'ovBalls/2', 'ovBalls/3', 'ovBalls/4',
       'ovBalls/5', 'ovBalls/6', 'id', 'inningId', 'ovBalls/7', 'ovBalls/8',
       'ovBalls/9', 'ovBalls/10', 'ovBalls/11', 'ovBalls/12', 'ovBalls/13',
       'bowlerId'],
      dtype='object')

In [ ]:
imputer = SimpleImputer(strategy='most_frequent')
sixBalls=overHistory.iloc[:,1:8]
overHistory[sixBalls.columns]=pd.DataFrame(imputer.fit_transform(columns),columns=sixBalls.columns)

In [ ]:
overHistory

ovNo ovBalls/0 ovBalls/1 ovBalls/2 ovBalls/3 ovBalls/4 ovBalls/5  \
0          1         1         .         .         .       1Wd         .   
1          2         1         .         .         .         1         1   
2          3         1         W         1         .       1Nb         .   
3          4         4       1Nb         .         .         .         4   
4          5         .         .         4         .         .         .   
...      ...       ...       ...       ...       ...       ...       ...   
110406    31         1         .         1         W         .         .   
110407    32         .         .         .         .         .         .   
110408    33         W         4         4         .         4         .   
110409    34         4         2         4         1         .         .   
110410    35         4         .         W         .         .         .   

       ovBalls/6      id  inningId ovBalls/7 ovBalls/8 ovBalls/9 ovBalls/10  \
0              1       1         1       NaN       NaN       NaN        NaN   
1              .       2         1       NaN       NaN       NaN        NaN   
2              .       3         1       NaN       NaN       NaN        NaN   
3              .       4         1       NaN       NaN       NaN        NaN   
4              .       5         1       NaN       NaN       NaN        NaN   
...          ...     ...       ...       ...       ...       ...        ...   
110406         .  110407      2682       NaN       NaN       NaN        NaN   
110407         .  110408      2682       NaN       NaN       NaN        NaN   
110408         .  110409      2682       NaN       NaN       NaN        NaN   
110409         .  110410      2682       NaN       NaN       NaN        NaN   
110410         .  110411      2682       NaN       NaN       NaN        NaN   

       ovBalls/11 ovBalls/12 ovBalls/13  bowlerId  
0             NaN        NaN        NaN       NaN  
1             NaN        NaN        NaN       NaN  
2             NaN        NaN        NaN       NaN  
3             NaN        NaN        NaN       NaN  
4             NaN        NaN        NaN       NaN  
...           ...        ...        ...       ...  
110406        NaN        NaN        NaN   4567.00  
110407        NaN        NaN        NaN   1894.00  
110408        NaN        NaN        NaN   4567.00  
110409        NaN        NaN        NaN   1894.00  
110410        NaN        NaN        NaN   4567.00  

[110411 rows x 18 columns]

The rest of the columns except bowlerId can be imputed with 0

In [ ]:
restBalls=overHistory.iloc[:,10:17]
overHistory[restBalls.columns]=overHistory.iloc[:,10:17].fillna(0)

The Bowler Id will be imputed as missing

In [ ]:
overHistory['bowlerId']=overHistory['bowlerId'].fillna('Missing_id')

In [ ]:
overHistory.isnull().sum()

ovNo          0
ovBalls/0     0
ovBalls/1     0
ovBalls/2     0
ovBalls/3     0
ovBalls/4     0
ovBalls/5     0
ovBalls/6     0
id            0
inningId      0
ovBalls/7     0
ovBalls/8     0
ovBalls/9     0
ovBalls/10    0
ovBalls/11    0
ovBalls/12    0
ovBalls/13    0
bowlerId      0
dtype: int64

# Players


In [ ]:
players.head()

id       fullName    shortName nationality dateOfBirth  rightArmedBowl  \
0  2065  Rick McCosker  RB McCosker  Australian  1946-12-11           False   
1  2114    Alan Turner     A Turner  Australian  1950-07-23           False   
2  2083  Greg Chappell   G Chappell  Australian  1948-08-07            True   
3  2115   Jeff Thomson    J Thomson  Australian  1950-08-16            True   
4  2020   Ian Chappell   I Chappell  Australian  1943-09-26            True   

   rightHandedBat bowlingStyle  teamId  
0            True          NaN    2808  
1            True          NaN    2808  
2            True  MEDIUM_SEAM    2808  
3            True    FAST_SEAM    2808  
4            True     LEG_SPIN    2808

In [ ]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28911 entries, 0 to 28910
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              28911 non-null  int64 
 1   fullName        28911 non-null  object
 2   shortName       28911 non-null  object
 3   nationality     28249 non-null  object
 4   dateOfBirth     28349 non-null  object
 5   rightArmedBowl  28911 non-null  bool  
 6   rightHandedBat  28911 non-null  bool  
 7   bowlingStyle    18328 non-null  object
 8   teamId          28911 non-null  int64 
dtypes: bool(2), int64(2), object(5)
memory usage: 1.6+ MB


In [ ]:
players.isnull().sum()

id                    0
fullName              0
shortName             0
nationality         662
dateOfBirth         562
rightArmedBowl        0
rightHandedBat        0
bowlingStyle      10583
teamId                0
dtype: int64

In [ ]:
players.describe(include='O')

fullName shortName nationality dateOfBirth bowlingStyle
count         28911     28911       28249       28349        18328
unique         1850      1865          50        1623            6
top     Virat Kohli   V Kohli      Indian  1987-04-30  MEDIUM_SEAM
freq            174       174        2851         184         5389

In [ ]:
koh=players['fullName']=='Virat Kohli'

In [ ]:
koh.sum()

174

In [ ]:
players.describe()

id        teamId
count   28911.00      28911.00
mean    11743.26   17282849.36
std     89965.08  152852594.54
min         1.00         11.00
25%       391.00       2997.00
50%      1075.00       3882.00
75%      2293.00      10223.50
max   1277627.00 2147212060.00